src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_NullablePublicOnly.cs

```diff
@@ -195,5 +195,154 @@ private static void AssertAttributes(ImmutableArray<CSharpAttributeData> attribu
             var actualNames = attributes.Select(a => a.AttributeClass.ToTestDisplayString()).ToArray();
             AssertEx.SetEqual(actualNames, expectedNames);
         }
+
+        [Fact]
+        [WorkItem(36457, "https://github.com/dotnet/roslyn/issues/36457")]
+        public void ExplicitAttribute_ReferencedInSource_Assembly()
+        {
+            var sourceAttribute =
+@"namespace System.Runtime.CompilerServices
+{
+    internal class NullablePublicOnlyAttribute : System.Attribute
+    {
+        internal NullablePublicOnlyAttribute(bool b) { }
+    }
+}";
+            var source =
+@"using System.Runtime.CompilerServices;
+[assembly: NullablePublicOnly(false)]
+[module: NullablePublicOnly(false)]
+";
+
+            // C#7
+            var comp = CreateCompilation(new[] { sourceAttribute, source }, parseOptions: TestOptions.Regular7);
+            verifyDiagnostics(comp);
+
+            // C#8
+            comp = CreateCompilation(new[] { sourceAttribute, source });
+            verifyDiagnostics(comp);
+
+            static void verifyDiagnostics(CSharpCompilation comp)
+            {
+                comp.VerifyDiagnostics(
+                // (3,10): error CS8335: Do not use 'System.Runtime.CompilerServices.NullablePublicOnlyAttribute'. This is reserved for compiler usage.
+                // [module: NullablePublicOnly(false)]
+                Diagnostic(ErrorCode.ERR_ExplicitReservedAttr, "NullablePublicOnly(false)").WithArguments("System.Runtime.CompilerServices.NullablePublicOnlyAttribute").WithLocation(3, 10));
+            }
+        }
+
+        [Fact]
+        [WorkItem(36457, "https://github.com/dotnet/roslyn/issues/36457")]
+        public void ExplicitAttribute_ReferencedInSource_Other()
+        {
+            var sourceAttribute =
+@"namespace System.Runtime.CompilerServices
+{
+    internal class NullablePublicOnlyAttribute : System.Attribute
+    {
+        internal NullablePublicOnlyAttribute(bool b) { }
+    }
+}";
+            var source =
+@"using System.Runtime.CompilerServices;
+[NullablePublicOnly(false)]
```

Comment:
Should we get an error on this attribute?

Reply:
Only if `AttributeTargets` is set.

---
In reply to: [296932751](https://github.com/dotnet/roslyn/pull/36152#discussion_r296932751) [](ancestors = 296932751)